colexecargs: fix recent memory leak

yuzefovich opened the pull request:
**colfetcher: set all unneeded vectors to all nulls**

The cFetcher creates a batch with the same schema as the table it is
reading from. In many cases not all columns from the table are needed
for the query, so as an important perfomance optimization the cFetcher
doesn't decode the data for those columns. As a result, such unneeded
columns are left "unset". This works ok in most cases, however, if we
attempt to materialize such a batch with unset vectors, the conversion
to datums might encounter errors (e.g. UUID values must either be NULL
or have 16 bytes length).

This commit improves this situation slightly by tracking the set of
unneeded columns and setting those vectors to all NULL values. This
will allow to simplify the planning code a bit in the follow-up commit.

Release note: None

**colbuilder: remove unnecessary complication when wrapping table reader**

Previously, we had some custom code for the case when we supported the
table reader core but not the post-processing spec - we attempted to
revert to the core-pre-planning state and plan the whole table reader
with render expressions on top.

Given the previous commit, I think this is no longer necessary, so this
commit removes that special code in favor of the general handling of
only the post-processing spec via a noop processor. This commit was
prompted by some complications because of this old code for the
follow-up commit.

Release note: None

**colexecargs: fix recent memory leak**

In c3b1617 we introduced a new utility
struct that keeps information about the meta objects in the operator
tree. Those meta objects are tracked by several slices which are
resliced to be of length 0 when the "outer" object is released back to
the corresponding pool. However, the slices still end up holding
references to the old meta objects prohibiting those from being GCed.
Such a behavior results in a memory leak. This commit fixes the issue by
explicitly resetting the slices for reuse.

Fixes: #62320.
Fixes: #64093.

Release note: None

**sql: audit implementations of Releasable interface of slices' reuse**

This commit performed the audit of all slices that are kept by
components implementing `execinfra.Releasable` interface to make sure
that the slices that might be referencing large objects are deeply
reset. (By deep reset I mean all slots are set to `nil` so that the
possibly large objects could be garbage-collected.) This was prompted by
the previous commit which fixed a recent regression, but this commit
seems like a good idea on its own, and it might be worth backporting it
too.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/64453)
<!-- Reviewable:end -->

yuzefovich committed:
colbuilder: remove unnecessary complication when wrapping table reader

Previously, we had some custom code for the case when we supported the
table reader core but not the post-processing spec - we attempted to
revert to the core-pre-planning state and plan the whole table reader
with render expressions on top.

Given the previous commit, I think this is no longer necessary, so this
commit removes that special code in favor of the general handling of
only the post-processing spec via a noop processor. This commit was
prompted by some complications because of this old code for the
follow-up commit.

Release note: None

yuzefovich committed:
colfetcher: set all unneeded vectors to all nulls

The cFetcher creates a batch with the same schema as the table it is
reading from. In many cases not all columns from the table are needed
for the query, so as an important perfomance optimization the cFetcher
doesn't decode the data for those columns. As a result, such unneeded
columns are left "unset". This works ok in most cases, however, if we
attempt to materialize such a batch with unset vectors, the conversion
to datums might encounter errors (e.g. UUID values must either be NULL
or have 16 bytes length).

This commit improves this situation slightly by tracking the set of
unneeded columns and setting those vectors to all NULL values. This
will allow to simplify the planning code a bit in the follow-up commit.

Release note: None

jordanlewis reviewed (commented):
The first 2 commits LGTM, 3rd I will review again a bit later

> Reviewed 2 of 2 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64453#-:-MZYzAK1-MQ1VQuRbmv8:b-hnlnm8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @yuzefovich)

___
*[pkg/sql/colfetcher/cfetcher.go, line 1499 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64453#-MZYnA1p59q-hBDlBS4f:-MZYnA1p59q-hBDlBS4g:b9s0rpk) ([raw file](https://github.com/cockroachdb/cockroach/blob/0cc7b91a014d84b4eef054dc8f07c90c0b80123c/pkg/sql/colfetcher/cfetcher.go#L1499)):*
> ```Go
> 	// We need to set all values in "not needed" vectors to nulls because if the
> 	// batch is materialized (i.e. values are converted to datums), the
> 	// conversion of unset values might encounter an error.
> ```

bummer... can we instead teach the materializer to notice this case?
___
*[pkg/sql/logictest/testdata/logic_test/vectorize_overloads, line 713 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64453#-MZYz2e89u-TWPHmQ4Yg:-MZYz2e89u-TWPHmQ4Yh:b-nkapli) ([raw file](https://github.com/cockroachdb/cockroach/blob/0cc7b91a014d84b4eef054dc8f07c90c0b80123c/pkg/sql/logictest/testdata/logic_test/vectorize_overloads#L713)):*
> ```
> └ Node 1
>   └ *rowexec.noopProcessor
>     └ *colfetcher.ColBatchScan
> ```

wrong commit i think


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64453)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis)

___
*[pkg/sql/colfetcher/cfetcher.go, line 1499 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64453#-MZYnA1p59q-hBDlBS4f:-MZYzeP7AxtY6e_wvpaz:b-mfk8lb) ([raw file](https://github.com/cockroachdb/cockroach/blob/0cc7b91a014d84b4eef054dc8f07c90c0b80123c/pkg/sql/colfetcher/cfetcher.go#L1499)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

bummer... can we instead teach the materializer to notice this case?
</blockquote></details>

It'll be a bit harder and less clean IMO - the materializer has the built-in assumption that all vectors need to be converted, and `ColBatchScan` is the only operator that might produce batches with unneeded vectors. We would need to examine the post-processing spec at the materializer level to learn which vectors are not needed, or we would have to plumb this information from the ColBatchScan somehow. This comes from the fact that we only the post-processing spec knows about projections and renders (TableReader core is an exception) that encoded this "unneediness" information.

I prefer the current approach because we're pushing this exceptional logic into the component that creates "exceptional" batches, that component already knows about not needed columns, and setting all unneeded vectors to all nulls works well for the conversion in the materializer.
___
*[pkg/sql/logictest/testdata/logic_test/vectorize_overloads, line 713 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/64453#-MZYz2e89u-TWPHmQ4Yg:-MZYzbjx9C0xiyxXXW8b:b-jqpf2h) ([raw file](https://github.com/cockroachdb/cockroach/blob/0cc7b91a014d84b4eef054dc8f07c90c0b80123c/pkg/sql/logictest/testdata/logic_test/vectorize_overloads#L713)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

wrong commit i think
</blockquote></details>

Yeah, I noticed it too - fixed.


<!-- Sent from Reviewable.io -->

yuzefovich committed:
colexecargs: fix recent memory leak

In c3b16171e2a886a9a92a1d824226d06e89fffeed we introduced a new utility
struct that keeps information about the meta objects in the operator
tree. Those meta objects are tracked by several slices which are
resliced to be of length 0 when the "outer" object is released back to
the corresponding pool. However, the slices still end up holding
references to the old meta objects prohibiting those from being GCed.
Such a behavior results in a memory leak. This commit fixes the issue by
explicitly resetting the slices for reuse.

Release note: None

yuzefovich commented:
Changed the third commit to a smaller, more proper fix, but I think keeping the first two commits is still worth it.

yuzefovich committed:
sql: audit implementations of Releasable interface of slices' reuse

This commit performed the audit of all slices that are kept by
components implementing `execinfra.Releasable` interface to make sure
that the slices that might be referencing large objects are deeply
reset. (By deep reset I mean all slots are set to `nil` so that the
possibly large objects could be garbage-collected.) This was prompted by
the previous commit which fixed a recent regression, but this commit
seems like a good idea on its own, and it might be worth backporting it
too.

Release note: None

yuzefovich commented:
Added another commit auditing our slice reuse in `Release` implementations as per discussion with @jordanlewis. I think a couple of places could be concerning, so that commit might be worthy of a backport.

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64453)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @yuzefovich)

___
*[pkg/sql/colconv/vec_to_datum_tmpl.go, line 103 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/64453#-MZoC2t-2Q1GZX_eAgTo:-MZoC2t-2Q1GZX_eAgTp:bkjh4tr) ([raw file](https://github.com/cockroachdb/cockroach/blob/b121d57686d0eae018a5952549fe40c014dca8fa/pkg/sql/colconv/vec_to_datum_tmpl.go#L103)):*
> ```Go
> 	for _, vec := range c.convertedVecs {
> 		if len(vec) > 0 {
> 			_ = vec[len(vec)-1]
> ```

What are we trying to do here? Why would `for i := range vec` have bound checks in the first place? By the way, I believe that `for i := range vec { vec[i] = nil }` has a fast path in Go and gets converted to a `memset`.
___
*[pkg/sql/colexec/colexecargs/op_creation.go, line 180 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/64453#-MZoCN2b1zM9Q0n5oMuz:-MZoCN2b1zM9Q0n5oMv-:bwsuwwk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b121d57686d0eae018a5952549fe40c014dca8fa/pkg/sql/colexec/colexecargs/op_creation.go#L180)):*
> ```Go
> 		// objects are still referenced by the corresponding sync.Pools, so the
> 		// references in r.Releasables will not be the reason for the objects to
> 		// not be garbage-collected.
> ```

Each sync.Pool grows and shrinks as necessary, it is possible that those other pools would shrink and leave this as the only reference.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64453)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @RaduBerinde)

___
*[pkg/sql/colconv/vec_to_datum_tmpl.go, line 103 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/64453#-MZoC2t-2Q1GZX_eAgTo:-MZp677cDDClD7yQvjTq:bt50kwj) ([raw file](https://github.com/cockroachdb/cockroach/blob/b121d57686d0eae018a5952549fe40c014dca8fa/pkg/sql/colconv/vec_to_datum_tmpl.go#L103)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

What are we trying to do here? Why would `for i := range vec` have bound checks in the first place? By the way, I believe that `for i := range vec { vec[i] = nil }` has a fast path in Go and gets converted to a `memset`.
</blockquote></details>

Good point - I was trying to eliminate bounds checks and blindly followed the way we usually do it in other places. The crucial difference is that here `vec` is a local variable, so the compiler can prove that the loop `for i := range vec` is always in bounds, whereas usually in other places the thing to iterate over is a field in the struct. I confirmed this here: https://godbolt.org/z/KevehETje. It does use `memclrHasPointers` which is likely what you were referring to as an optimization.

(A side thought is that possibly in other places we should follow the same pattern of declaring a local variable to access the field in the struct, and then iterate over the local variable rather than our somewhat ugly `[len(s)-1]` accesses outside of the for loop.)

Refactored.
___
*[pkg/sql/colexec/colexecargs/op_creation.go, line 180 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/64453#-MZoCN2b1zM9Q0n5oMuz:-MZp7WY-9H0bL4PPmPxc:bsg2nw) ([raw file](https://github.com/cockroachdb/cockroach/blob/b121d57686d0eae018a5952549fe40c014dca8fa/pkg/sql/colexec/colexecargs/op_creation.go#L180)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Each sync.Pool grows and shrinks as necessary, it is possible that those other pools would shrink and leave this as the only reference.
</blockquote></details>

Makes sense, fixed.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
@jordanlewis @RaduBerinde I believe that both of are on board with this PR, can someone give it another look and stamp if satisfied?

jordanlewis reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64453#-:-MZtxxEn8bN3WLKfvABA:b-uwkowd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis, @RaduBerinde, and @yuzefovich)

___
*[pkg/sql/colflow/vectorized_flow.go, line 1264 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/64453#-MZtxr-76ulMK8-TiVBP:-MZtxr-82mHgh4IIl-RT:bvqkkmr) ([raw file](https://github.com/cockroachdb/cockroach/blob/5961f6238fbd7528f077bf2e25192e1a5764bfc9/pkg/sql/colflow/vectorized_flow.go#L1264)):*
> ```Go
> 	// slice). Unset the slot so that we don't keep the reference to the old
> 	// materializer.
> 	if len(r.processors) == 1 {
> ```

If this comment is true, why do we keep this as a slice at all? Should it just be a pointer?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64453#-:-MZtzcSMBoM6zEDoACfE:b-tlx7ko)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis and @RaduBerinde)

___
*[pkg/sql/colflow/vectorized_flow.go, line 1264 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/64453#-MZtxr-76ulMK8-TiVBP:-MZtzKawEeXt0ukveL5_:b21rocd) ([raw file](https://github.com/cockroachdb/cockroach/blob/5961f6238fbd7528f077bf2e25192e1a5764bfc9/pkg/sql/colflow/vectorized_flow.go#L1264)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

If this comment is true, why do we keep this as a slice at all? Should it just be a pointer?
</blockquote></details>

It's because in the row-based flows can have multiple processors, so we abstracted out `SetProcessors` method that takes in a pointer. We do pool the slice as part of `vectorizedFlowCreatorHelper` pooling and instantiate it with 1 capacity.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Rebased on top of master.

bors r+
